# <center> Problem Set 4

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import t
from scipy.optimize import minimize
from scipy import stats
from scipy.optimize import fmin_bfgs
import math
import statsmodels.api as sm

df = pd.read_stata("PS4_data.dta")
#df.describe()

### 1. Select only male heads of household who are between 25 and 60 years of age and earn wages > $7/hr

In [2]:
#create wage variable using income and work hours
#df['wage'].describe()
df['wage2'] = df['hlabinc']/df['hannhrs']
df = df[(df.hsex==1)&(df.age>=25)&(df.age<=60)&(df.wage2>7)]

#create log wage variable
df['lnwage2'] = np.log(df['wage2'])
df = df[df.lnwage2!=np.inf]

#create age squared variable
df['agesq'] = df['age']**2

#df.agg({"age": ["mean", "min", "max"], "hsex": ["mean", "min", "max"]})
sumstats=df[['hsex','age','agesq','wage2','lnwage2']]
sumstats.describe().loc[['count','mean','max','min']]

hsex           age         agesq         wage2       lnwage2
count  57477.0  57477.000000  57477.000000  57477.000000  57477.000000
mean       1.0     39.224247   1629.789673     24.306034      3.010414
max        1.0     60.000000   3600.000000   1717.330322      7.448526
min        1.0     25.000000    625.000000      7.000252      1.945946

### 2. Create indicator and continuous variables as necessary

In [3]:
#df['hrace'].describe()
df['White'] = (df['hrace'] == 1).astype(int)
df['Black'] = (df['hrace'] == 2).astype(int)
df['Hispanic'] = (df['hrace'] == 5).astype(int)
df['OtherRace'] = ((df['hrace']!=1)&(df['hrace']!=2)&(df['hrace']!=5)).astype(int)

#create df for the model
df = df[['lnwage2','hyrsed', 'age', 'agesq', 'Black', 'Hispanic', 'OtherRace', 'White', 'year']].dropna()

#df.agg({"age": ["mean", "min", "max"], "hsex": ["mean", "min", "max"]})
sumstats=df[['Black', 'Hispanic', 'OtherRace', 'White']]
sumstats.describe().loc[['count','mean','max','min']]

Black  Hispanic     OtherRace         White
count  57097.000000   57097.0  57097.000000  57097.000000
mean       0.056343       0.0      0.023119      0.920539
max        1.000000       0.0      1.000000      1.000000
min        0.000000       0.0      0.000000      0.000000

### 3. Estimate the following model via a Maximum Likelihood Estimator separately for t = 1971, 1980, 1990, 2000

#### Create Linear MLE Function

Nomral Log-Likelihood function:
\begin{equation}
\begin{gathered}
\ln \mathcal{L}(\theta \mid y, x)=-\frac{n}{2} \ln (2 \pi)-\frac{n}{2} \ln \left(\sigma^{2}\right)- \frac{1}{2 \sigma^{2}}(y-x \beta)^{\prime}(y-x \beta)
\end{gathered}
\end{equation}

In [4]:
class Linear_Model_MLE(object):
    '''
    This is a Class for Linear multiple regression model using MLE estimation
    The Optimaization method is BFGS, but you can modify the code and use other methods
    
    Function Linear_MLE: calculate the values for coefficent estimates, SE, variance covariance matrix, t-stats, and 
    p-value
    
    Function Summary: return regression results
                      For example: If you would like to see the regression summary table, 
                                   please write it like this example: Linear_Model_MLE(y,x, create_intercept=True).Summary()
    
    The class automatically delete variables that have only zero values.
    It is because if a variable has all zero values, the matrix would be singlular. 
    Please still make sure to use your variable only if it has relatively large enough non-zero observation 
    because the model would lose degree of freedom as you add more variables.
   
   Inputs:
    y: dependent variable (Type: pd Series)
    x: independent variable (Type: pd DataFrame, or pd Series if there is only one x variable)
    create_intercept: True or False (Type: bool). The default is True. Alter based on your model. 
    '''
    
    def __init__(self, y, x, create_intercept=True):
        self.y = y
        #self.x = x
        self.x = x.loc[:, (df != 0).any(axis=0)] #drop if variable is all zero

        if create_intercept:
            self.x['intercept']=1
            #self.x = self.x.assign(intercept=pd.Series([1]*np.shape(self.x)[0]))
            self.x = self.x
        else:
            self.x = x

        if isinstance(create_intercept, bool):
            self.create_intercept = create_intercept
        else:
            raise RuntimeError("error")        

    def Linear_MLE(self):
        self.xt = self.x.T
        self.xx = self.xt @ (self.x)
        self.xxi = np.linalg.inv(self.xx)

        n = self.x.shape[0] #nrow
        k = self.x.shape[1] #ncol
        GuessVector = np.ones(k+1) # beta Vector is initial guess for beta value                
        
        def neglnL(theta):
            beta=theta[:-1]  #everything but the last item
            sigma=theta[-1:] #the last item
            yhat = self.x@beta
            
            lnL = -(n/2)*np.log(2*math.pi) - (n/2)*np.log(sigma**2) - (1/(2*sigma **2)) * (self.y-yhat)@(self.y-yhat).T
            #lnL = np.sum(stats.norm.logpdf(y, loc=yhat, scale=sigma)) #method 2
            return -lnL
          
        estimates = minimize(neglnL, GuessVector, method='BFGS')
        #estimates = fmin_bfgs(neglnL, x0=GuessVector, disp=False)
        
        bhat = estimates.x[:-1]
        var = (1/n) * (self.y- self.x@bhat) @ (self.y- self.x@bhat).T
        vcv = var * self.xxi
        standard_error = np.sqrt(np.diag(vcv))
        z_stat = bhat/standard_error
        p_value = t.sf(np.abs(z_stat), n-k)*2  

        self.results = {"Variable name": self.x.columns.values,
                        "coefficient": bhat, 
                        "standard_error": standard_error, 
                        "z_stat": z_stat, 
                        "p_value": p_value, 
                        "create_intercept": self.create_intercept}
     
    def summary(self):
        self.Linear_MLE()        
        print("labels","coefficient value","standard error","t-statistic","p-value",sep="\t    ")
        for i in range(len(self.results['Variable name'])):
            results11 = str(self.results['Variable name'][i]) + '\t' + '\t     ' + str(round(self.results['coefficient'][i], 4)) + '\t' + '\t' + '\t'  + str(round(self.results['standard_error'][i], 4)) + '\t              '  + str(round(self.results['z_stat'][i], 4)) + '\t    ' + str(round(self.results['p_value'][i], 4))
            print(results11)

### The model

$\ln \left(w_{i, t}\right)=\alpha+\beta_{1} E d u c_{i, t}+\beta_{2}$ Age $_{i, t}+\beta_{3} A g e_{i, t}^{2}+\beta_{4}$ Black $_{i, t}+\beta_{5}$ Hispanic $_{i, t}+\beta_{6}$ OtherRace $_{i, t}+\varepsilon_{i, t}$
w
where:
- $w_{i, t}=$ wage of individual $i$ in survey year $t$
- $E d u c_{i, t}=$ education in years
- $A g e_{i, t}=$ age in years
- Black$_{i, t}$, Hispanic$_{i, t}$, OtherRace$_{i, t}=$ dummy variables for race $=$ Black, Hispanic, Not $\in\{$ White, Black, Hispanic $\}$.

#### Estimate the model if t = 1971

In [5]:
df71 = df[df.year == 1971]
x = df71[['hyrsed', 'age', 'agesq', 'Black', 'Hispanic', 'OtherRace']]
y = df71['lnwage2']
Fit = Linear_Model_MLE(y,x, create_intercept=True)
Fit.summary()

labels	    coefficient value	    standard error	    t-statistic	    p-value
hyrsed		     0.0665			0.0037	              17.929	    0.0
age		     0.0647			0.0094	              6.847	    0.0
agesq		     -0.0006			0.0001	              -5.3594	    0.0
Black		     -0.1642			0.0445	              -3.6891	    0.0002
OtherRace		     0.0176			0.0678	              0.2591	    0.7956
intercept		     0.5911			0.1929	              3.0645	    0.0022


In [6]:
#sm_OLS_fit = sm.OLS(endog=y, exog=x).fit()
#sm_OLS_results= sm_OLS_fit.summary()
#print(sm_OLS_results)

#### Estimate the model if t = 1980

In [7]:
df80 = df[df.year == 1980]
x = df80[['hyrsed', 'age', 'agesq', 'Black', 'Hispanic', 'OtherRace']]
y = df80['lnwage2']
Fit = Linear_Model_MLE(y,x, create_intercept=True)
Fit.summary()

labels	    coefficient value	    standard error	    t-statistic	    p-value
hyrsed		     0.066			0.0043	              15.4805	    0.0
age		     0.0455			0.0096	              4.7594	    0.0
agesq		     -0.0004			0.0001	              -3.4495	    0.0006
Black		     -0.103			0.0434	              -2.3721	    0.0178
OtherRace		     0.0126			0.0708	              0.178	    0.8588
intercept		     1.0044			0.1915	              5.2456	    0.0


In [8]:
#sm_OLS_fit = sm.OLS(endog=y, exog=x).fit()
#sm_OLS_results= sm_OLS_fit.summary()
#print(sm_OLS_results)

#### Estimate the model if t = 1990

In [9]:
df90 = df[df.year == 1990]
x = df90[['hyrsed', 'age', 'agesq', 'Black', 'Hispanic', 'OtherRace']]
y = df90['lnwage2']
Fit = Linear_Model_MLE(y,x, create_intercept=True)
Fit.summary()

labels	    coefficient value	    standard error	    t-statistic	    p-value
hyrsed		     0.0955			0.0049	              19.5832	    0.0
age		     0.0576			0.0106	              5.4166	    0.0
agesq		     -0.0005			0.0001	              -4.1786	    0.0
Black		     -0.1681			0.0475	              -3.5371	    0.0004
OtherRace		     -0.052			0.0886	              -0.587	    0.5572
intercept		     0.2825			0.2168	              1.3028	    0.1928


In [10]:
#sm_OLS_fit = sm.OLS(endog=y, exog=x).fit()
#sm_OLS_results= sm_OLS_fit.summary()
#print(sm_OLS_results)

#### Estimate the model if t = 2000

In [11]:
df2000 = df[df.year == 2000]
x = df2000[['hyrsed', 'age', 'agesq', 'Black', 'Hispanic', 'OtherRace']]
y = df2000['lnwage2']
Fit = Linear_Model_MLE(y,x, create_intercept=True)
Fit.summary()

labels	    coefficient value	    standard error	    t-statistic	    p-value
hyrsed		     0.1103			0.0051	              21.6434	    0.0
age		     0.0839			0.0102	              8.2469	    0.0
agesq		     -0.0009			0.0001	              -7.2132	    0.0
Black		     -0.2597			0.0473	              -5.4849	    0.0
OtherRace		     -0.0553			0.0546	              -1.0138	    0.3108
intercept		     -0.2838			0.2156	              -1.3159	    0.1883


#### Compare the MLE results against an StatsModel OLS estimator to confirm the results

In [12]:
X=sm.add_constant(x)
sm_OLS_fit = sm.OLS(endog=y, exog=X).fit()
sm_OLS_results= sm_OLS_fit.summary()
print(sm_OLS_results)

                            OLS Regression Results                            
Dep. Variable:                lnwage2   R-squared:                       0.222
Model:                            OLS   Adj. R-squared:                  0.221
Method:                 Least Squares   F-statistic:                     148.1
Date:                Mon, 04 Oct 2021   Prob (F-statistic):          1.41e-138
Time:                        12:21:25   Log-Likelihood:                -2055.3
No. Observations:                2595   AIC:                             4123.
Df Residuals:                    2589   BIC:                             4158.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.2902      0.216     -1.344      0.1

C:\Users\haimiti.aerfate\Softwares\anaconda3\lib\site-packages\statsmodels\base\model.py:1362: RuntimeWarning: invalid value encountered in true_divide
  return self.params / self.bse
C:\Users\haimiti.aerfate\Softwares\anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:1932: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= _a)


#### MLE vs OLS

My MLE model results are very similar to the StatsModel OLS results in terms of coefficients, SE, t-statistics, and p-value. I only showed the StatsModel OLS results for 2000. If you would like to see the StatsModel OLS result for another year, please uncomment the code provided, and run the code.

### 4. Interpret the coefficient β1. How do the returns to education change over time in these data?

Since the Hispanic variable has only zero values, the regression result does not include Hispanic. I made my MLE model to delete a variable if it only has all zero values. It is because if there a all zero variable, the matrix would be singlular. 

The model suggests that all else equal, the positive effect of education on wages has increased over the years. In other words, the wage between the less educated people and highly educated people has increased. For example, the models suggest in 1971 and 1980, one more year of education lead to around a 0.066 dollar increase in wages. However, in 1990 and 2000, the effect has increased to around 0.096 and 0.11 dollars respectively. The wage variable is statistically significant in all models at a 1% level of significance. The wages in the model are deflated to 2005.